datafusion/catalog/src/table.rs

```diff
@@ -247,6 +247,9 @@ pub trait TableProvider: Debug + Sync + Send {
     }
 
     /// Get statistics for this table, if available
+    /// Although not presently used in mainline DataFusion, this allows implementation specific
+    /// behavior for downstream repositories, in conjunction with specialized optimizer rules to
+    /// perform operations such as re-ordering of joins.
```

Comment:
Why don't we add such optimizer in DataFusion core, @alamb ?

Reply:
I think the theory was that the ExecutionPlan (aka physical plan) had access to more specific information / statistics (like which files would be read, min/max values, etc) so doing join reordering could produce better results when done at that level

This is why the current join selection is a physical optimizer pass https://github.com/apache/datafusion/blob/main/datafusion/core/src/physical_optimizer/join_selection.rs

However, for systems that have a good catalog / some other source of statistical information, I could imagine doing the reordering at the logical level

Reply:
I agree this being the case for ParquetExec / AvroExec etc where stats come from file stats.
I see there isn't an ExecutionPlan which would generically wrap a TableProvider... Not sure why we have this method then, if it cannot be used directly in DF. 

> However, for systems that have a good catalog / some other source of statistical information, I could imagine doing the reordering at the logical level

 in fact having a metastore/catalog should not be uncommon. Many data landscapes center around a catalog, eg when Iceberg is involved

Reply:
> Not sure why we have this method then, if it cannot be used directly in DF.

It can still be used in logical optimizer rules.

Reply:
Maybe we could write up some example in `datafusion-examples` that show how to use it / make it clearer how to do so